HFStripFilter for Run 2 data

toropin opened the pull request:
#### PR description:

This PR is devoted to the development of a new filter HFStripFilter
for HF RecHits. This work was started when the JetMET group found a
lot of FakJets originated in HF. A careful examination revealed that
the high-energy hits from these jets have times ~4-6 ns and Ieta
mostly 29-34. This indicated that these hits could be the residual
tails after TDC cuts with QIE10. In Ieta 29-34 the distribution of
signals in time corresponding to early hits caused by possible PMT
hits are not well separated from the main physical signals and hits
with ~4-6 ns are probably due to tails from PMT hits (See slide 13
in ref. 1).
These early hits may be due to muons passing through "jungle" of
optical fibers coming from the HF absorbers to PMTs. In this case, we
can have a sequence of hits in one HF wedge, which can have signals in
both depths 1 and 2 and have a strip-like structure: sequence of hits
in one wedge and without or with very small signals in adjacent
wedges. Such signals cannot be marked with other HF filters, because
signals basically have comparable hits in both depths 1 and 2 and PET
and S9S1 HF filters work with hits which have very large signal in one
of depths and very small signal or no signal in another depth.
So HFStripFilter is looking for strips in HFRecHitCollection and marks
hits in the strips if they were found with special flag which could be
used for further data processing (See first part of report in ref. 2).

References:
1. https://indico.cern.ch/event/800247/contributions/3325296/attachments/1800754/2937107/HFStripFilter_22Feb2019.pdf
2. https://indico.cern.ch/event/807827/contributions/3362430/attachments/1816769/2969579/HFStripFilter_22Mar2019.pdf

#### PR validation:
The logs produced by "runTheMatrix" script look reasonable. There are few errors in log files, but one error related  to data server  eoscms.cern.ch which appears to be down or unreachable (cmsdev31.cern.ch). Some other workflows with errors have DAS_ERROR status which is also unrelated to this PR.

In attached Screenshots you can see pictures of EvenrDisplay (FireWorks) with HF rechits after reconstruction procedure. In column 'flags' you can see the upper 2 hits in table belonging to strip have values 1048640 (set bits 6, 20). Bit6 is new flag HcalPhase1FlagLabels::HFAnomalousHit). 
<img width="1201" alt="Screenshot 2019-03-25 at 17 21 55" src="https://user-images.githubusercontent.com/7665869/54983371-f8e16100-4fac-11e9-897c-c391d8a166ee.png">
<img width="1200" alt="Screenshot 2019-03-25 at 17 18 58" src="https://user-images.githubusercontent.com/7665869/54983400-04cd2300-4fad-11e9-9bbd-3daa3fc71d45.png">

igv4321 committed:
Framework for implementing HFStripFilter in CMSSW

igv4321 committed:
changing the "runFilter" signature

igv4321 committed:
added a parameter

toropin committed:
New method runFilter

toropin committed:
Removed unused parameters

toropin committed:
Removed unused parameters

toropin committed:
HFStripFilter: Removed unused parameters

toropin committed:
HFStripFilter: Removed unused parameters

toropin committed:
Modified debuggung printout

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8909

- This PR adds an extra 40KB to repository

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8909/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8909/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

toropin committed:
changed (.size() == 0) to (.empty()) to check vector size

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8917

- This PR adds an extra 36KB to repository

cmsbuild commented:
A new Pull Request was created by  @toropin  for master.

It involves the following packages:

DataFormats/METReco
RecoLocalCalo/HcalRecAlgos
RecoLocalCalo/HcalRecProducers


@cmsbuild, @perrotta, @slava77 can you please review it and eventually sign? Thanks.
@TaiSakuma, @gouskos, @ahinzmann, @rappoccio, @mmarionncern, @rovere, @argiro, @jdamgov, @jdolen, @nhanvtran, @gkasieczka, @clelange, @schoef, @mariadalfonso, @seemasharmafnal this is something you requested to watch as well.
@davidlange6, @slava77, @fabiocos you are the release manager for this.

cms-bot commands are listed <a href="http://cms-sw.github.io/cms-bot-cmssw-cmds.html">here</a>

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

perrotta commented:
Please @toropin : change the PR title with something more explanatory

cmsbuild commented:
-1

Tested at: 1dd98d339b8fefc610d874023e2946c0bb68a13e

You can see the results of the tests here:
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33766/summary.html

I found follow errors while testing this PR

**Failed tests:  ClangBuild**

* **Clang**:

I found compilation warning while trying to compile with clang. Command used:
```
USER_CUDA_FLAGS='--expt-relaxed-constexpr' USER_CXXFLAGS='-Wno-register -fsyntax-only' scram build -k -j 32 COMPILER='llvm compile'

```
 See details on the summary page.

cmsbuild commented:
Comparison not run due to Build errors/Fireworks only changes/No short matrix requested (RelVals and Igprof tests were also skipped)

toropin committed:
changed fabs to abs for some cases

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8922

- This PR adds an extra 36KB to repository

cmsbuild commented:
Pull request #26251 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

perrotta reviewed (commented):
Just a few things noticed at a quick glance: they could be implemented before relaunching the tests

toropin committed:
Configurable parameters changed to Untracked, iterators changed to auto& it

toropin committed:
Configurable parameters changed to Untracked

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8933

- This PR adds an extra 76KB to repository

cmsbuild commented:
Pull request #26251 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

toropin committed:
Set some configurable parameters back to tracked type

toropin committed:
Set some configurable parameters back to tracked type

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8942

- This PR adds an extra 76KB to repository

cmsbuild commented:
Pull request #26251 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: d9d3909fbb32187ee58521da3ade4a1febeb8de8
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33795/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33795/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3114829
 - DQMHistoTests: Total failures: 5
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3114627
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,350 @@
+      if (it.time() > timeMax_ || it.time() < 0 || it.energy() < stripThreshold_) continue;	
+      // find HF hit with maximum signal in depth = 1
+      if (it.id().depth() == 1) {
+	if (it.energy() > d1max.energy() && std::abs(it.id().ieta()) < 35) {
```
Can this "35" be taken from somewhere, instead of being hardcoded here?

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,350 @@
+  }
+  
+  int phiseg = 2; // 10 degrees segmentation for most of HF (1 iphi unit = 5 degrees)
+  if (std::abs(d1strip[0].id().ieta()) > 39) phiseg = 4; // 20 degrees segmentation for |ieta| > 39
```
Also this "39" can probably be taken from some geometry: isn't it?

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,350 @@
+  
+  // Check if seed hit has neighbours with (iphi +/- phiseg) and the same ieta    
+  int iphi1 = d1strip[0].id().iphi() - phiseg;
+  while (iphi1 < 0) iphi1 += 72;
```
Same for "72" here

igv4321 commented on DataFormats/METReco/interface/HcalPhase1FlagLabels.h:
```diff
@@ -20,7 +20,8 @@ namespace HcalPhase1FlagLabels
         /*  3 */ HFS8S1Ratio = HcalCaloFlagLabels::HFS8S1Ratio,
         /*  4 */ HFPET = HcalCaloFlagLabels::HFPET,
-        /*  5 */ HFSignalAsymmetry = 5
+        /*  5 */ HFSignalAsymmetry = 5,
```
We keep HcalCaloFlagLabels untouched in case somebody wants to reprocess pre-Phase-1 data. The hardware was different then (no PMTs with dual anode readout, no hfprereco module in the sequence, etc), and the meaning of these bits was different as well. In particular, we had HSCP_ExpFit=5 and HBHETimingTrustBits=6. For Phase-1, the HCAL Noise Group performed a review of all status bits and decided that some of them can be repurposed.

toropin commented:
1. Changed definition of ieta sign by using signbit() from "math.h".

2. Replaced 'if (stripIphiMax == 0)' to 'if (d1max.energy() <= 0)'. It is
the same because it means strip was not found in depth = 1 and going to
use strip from depth = 2 to define wedge parameters.

3. Added configurable parameter seetHitIetaMax to replace hardcoded
value 35.

4. To define validity of values iphi = iphi +/- phiseg was copied
procedure from HcalHF_S9S1algorithm.cc. Replaced to methods incIPhi(...)
and decIPhi(...) from HcalTopology.

toropin committed:
Changed definition of ieta sign with signbit(). Changed definition of validity iphi = iphi +/- phiseg to using methods from HcalTopology.

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
-code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8999

- This PR adds an extra 80KB to repository

Code check has found code style and quality issues which could be resolved by applying a patch in https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8999/git-diff.patch
e.g. `curl https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/8999/git-diff.patch | patch -p1`

You can run `scram build code-checks` to apply code checks directly

perrotta commented:
please test

toropin committed:
<math.h> replaced by <cmath>

toropin committed:
additional protection for the possible case if d1strip has no hits

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/9001

- This PR adds an extra 92KB to repository

cmsbuild commented:
Pull request #26251 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: ce9f4a8a5d2d3df392014682fecd8fbaaaaf7294
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33876/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33876/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3139747
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3139549
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,361 @@
+  int stripDepthMax = 0;
+  
+  if (d1max.energy() > 0) {
+    signStripIeta = std::signbit(d1max.id().ieta());
```
std::signbit(arg) returns a bool, which is "false" if arg is positive, and "true" if arg is negative
While the logic here is correct, perhaps it could be more convenient to define all those local variables as "bool"

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,361 @@
+    desc.add<double>("timeMax", 6.0);
+    desc.add<double>("maxStripTime", 10.0);
+    desc.add<double>("wedgeCut", 0.05);
+    desc.add<int>("seedHitIetaMax", 35);
```
When I asked whether this value could have been taken from somewhere else instead of hardcoding it, I had in mind somewhere in the geometry, rather than a configuration parameter here.

I expect that the eta/phi segmentation must be defined somewhere, and it would be preferable that you retrieve the details of that segmentation (e.g. ietaMax) from such a central location, so that if anything get changed centrally it can automatically propagate everywhere else.

Defining it with a configuration parameter inside a given method that needs to know that parameter can make sense if that parameter in this method can also take different values with respect to the central ones. Therefore, if this "35" refers to the fact that "the high-energy hits from these jets have ... Ieta mostly 29-34", having that "35" defined here does make sense. Otherwise, from the practical point of view, it is not that different than hardcoding it in the c++.

igv4321 replied:
Alexander, please check if something in the HcalTopology class (Geometry/​CaloTopology/​interface/​HcalTopology.h) corresponds to the magic number 35. If yes, we can forward a reference to the HcalTopology instance to your code. If not, I suggest leaving the current code as it is.

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,361 @@
+      neigh = myqual->topo()->incIPhi(id, neighbour);
+      iphi2 = (neigh) ? neighbour.iphi() : 0;
+      if (it.id().iphi() == iphi1) energyIphi1 += it.energy();      // Energy iphi1
+      else if (it.id().iphi() == iphi2) energyIphi2 += it.energy(); // Energy iphi2
```
(Just to check the logic): is it really what you want to add up the energy in only one of the two neighbours?

perrotta commented on RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc:
```diff
@@ -0,0 +1,361 @@
+      if (it.time() > timeMax_ || it.time() < 0 || it.energy() < stripThreshold_) continue;	
+      // find HFhit with maximum signal in depth = 2
+      if (it.id().depth() == 2 && it.energy() > d2max.energy() && std::abs(it.id().ieta()) 
+          < seedHitIetaMax_) {
```
Please don't break the line here, for a better readability

perrotta commented:
Three more reduced HF recHits are present in miniAOD in TTbar13TeVPUppmx2018 wf 250202.181 (see jenkins test results): is this expected? Shouldn't a filter remove hits, instead of adding a few more?

igv4321 commented:
Yes, this is expected. The filter only adds a status bit to the rechits, but the "severity level" of this bit is set below the rechit elimination threshold. The final decision will thus be taken by the user. The reason is that the pattern recognition strategy used by the filter is not fool-proof, and it can actually tag valid rechits. Therefore, we have the usual efficiency vs. fake rate trade-off issues which can mean different things for different physics analyses.

perrotta commented:
> Yes, this is expected. The filter only adds a status bit to the rechits, but the "severity level" of this bit is set below the rechit elimination threshold. The final decision will thus be taken by the user. The reason is that the pattern recognition strategy used by the filter is not fool-proof, and it can actually tag valid rechits. Therefore, we have the usual efficiency vs. fake rate trade-off issues which can mean different things for different physics analyses.

I understand that they are not cut away, but then why do we have additional HF recHits instead? Am I misinterpreting the plots?

igv4321 commented:
I believe that the reduced HF recHit collection picks up all rechits with any status bit set.

perrotta commented:
The extra HF recHits appeared in the previous tests [1] in correspondance of a different workflow (TTbar13TeVPU2017 wf 10224.0 instead of TTbar13TeVPUppmx2018 wf 250202.181).

The non-reproducibility is in the HF recHits, which is suspiciously the same object tousched by this PR. However, this PR really only adds a single bit to the HFFlagMask, and it is hard to believed that it can be responsible of such a non-reproducibility.

Puzzled...

[1] https://github.com/cms-sw/cmssw/pull/26251#issuecomment-477212136

abdoulline commented:
Yes, indeed.


On Wed, 3 Apr 2019, Igor Volobouev wrote:

> 
> I believe that the reduced HF recHit collection picks up all rechits with any status
> bit set.
> 
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub, or mute the
> thread.[AEx02iHbNS2SRAh6FR5XwudbTzT5A_SFks5vdNysgaJpZM4cK7C6.gif]
> 
> 
>

toropin committed:
Changed type of singStripIeta to bool. Added some cosmetic changes in code.

toropin commented:
1. Changed 'int signStripIeta' to 'bool signStripIeta'. It should not
affect performance.

2. The values 72 and 39 were indeed the HF Topological parameters. The
value 35 is an experimentally defined value (see for example reference
https://indico.cern.ch/event/807827/contributions/3362430/attachments/1816769/2969579/HFStripFilter_22Mar2019.pdf
the rightmost plot in upper row of slide 6). It could be configurable
with possibility to make selection of seed hits more tightly.

3. Iphi is the indicator of wedge were strip is located. Iphi1 = Iphi
+ phiseg and Iphi2 = Iphi - phiseg (phiseg is phi segmentation) are
indicating the 2 neighbouring wedges. The energies in these wedges are
used to calculate the possible leak of energy from strip (isolation of
strip). Actually the sum of energies in the neighbouring wedges could
be used but wedgeCut parameter was not adjusted for this case.

4. The performance of few strings of code were improved.

5. It still reports warning during processing of command
'scram build code-checks' which I cannot understand.
clang-tidy -export-fixes tmp/slc7_amd64_gcc700/code-checks/RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc.yaml -header-filter '/afs/cern.ch/work/t/toropin/Phase1/CMSSW_10_6_X_2019-03-25-2300/src/.*' src/RecoLocalCalo/HcalRecAlgos/src/HFStripFilter.cc
735 warnings generated.
warning: argument unused during compilation: '--param vect-max-version-for-alias-checks=50' [clang-diagnostic-unused-command-line-argument]
Suppressed 734 warnings (733 in non-user code, 1 with check filters).

cmsbuild commented:
The code-checks are being triggered in jenkins.

cmsbuild commented:
+code-checks

Logs: https://cmssdt.cern.ch/SDT/code-checks/cms-sw-PR-26251/9069

- This PR adds an extra 92KB to repository

cmsbuild commented:
Pull request #26251 was updated. @cmsbuild, @perrotta, @slava77 can you please check and sign again.

perrotta commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: e8b16e00fda3f089fb04002478d573eedf985f37
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33955/summary.html

cmsbuild commented:
Comparison job queued.

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-26251/33955/summary.html

Comparison Summary:
 - No significant changes to the logs found
 - Reco comparison results: 5 differences found in the comparisons
 - DQMHistoTests: Total files compared: 32
 - DQMHistoTests: Total histograms compared: 3139747
 - DQMHistoTests: Total failures: 1
 - DQMHistoTests: Total nulls: 0
 - DQMHistoTests: Total successes: 3139549
 - DQMHistoTests: Total skipped: 197
 - DQMHistoTests: Total Missing objects: 0
 - DQMHistoSizes: Histogram memory added: 0.0 KiB( 31 files compared)
 - Checked 133 log files, 14 edm output root files, 32 DQM output files

perrotta commented:
Thank you @toropin @igv4321 @abdoulline for your updates and explanations.

Ok, now it is clear to me why there can be extra HF recHits if there is a status bit newly set.
What I still don't fully understand is why the location in which those additional recHits appear changed during the different tests made for this PR (at the beginning it was in wf 25020 and later it moved to wf  10224).

It could be that the input events changed already in the baseline, and therefore it can be that the filter acting on different events give some different result. Unfortunately, the matrix logs of the IB used for the first test got erased in the meanwhile, and this cannot be checked; the changes in the next two tests reproduce as expected, instead.

Just to be sure that nothing unwanted got included during the code updates that followed the review, could you please re-evaluate the performance with the latest version of this PR and confirm that they are still as expected?

toropin commented:
Actually after each new 'git push ...' I made full check using data file I have in my directory (events selected by JetMET group). These events are enriched with strip-like cases.
Now I did this check one more time. Everything looks the same as it was at the beginning of this PR.
So I can say the performance has not changed.

perrotta commented:
+1
- The status bit for the new HFStripFilter is set, and made available for possible evaluation
-  Extra HF reduced recHits can be saved if there is such a status bit newly set: this happens at a succesfully low rate that the increase in the event size is negligible
- Also the cpu time taken by this filter is below the resolution of the measurement errors, as verified with wf 11024
- Jenkins tests pass and the only differences reported are about those very few possible extra hfReducedRecHits

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next master IBs (tests are also fine). This pull request will now be reviewed by the release team before it's merged. @davidlange6, @slava77, @smuzaffar, @fabiocos (and backports should be raised in the release meeting by the corresponding L2)

fabiocos commented:
+1
